### ISSUE TITLE
New ModelKind: `EquationalRealm`

### ISSUE THREAD

**Issue by balacij** [JacquesCarette/Drasil#2440]
> I agree with the reasoning.
>
> We could indeed use an `NEList` of expressions (for the right hand sides) in `EquationalModels` to collapse the two together. But, let's not do that now. Using something from a list requires a choice, but in the 1 case, there is none - and from a design perspective, that's quite different.
>
>I'm thinking `EquationalRealm`. We use `Realm` in a paper to mean a theory with multiple interfaces.  Let's start with that?

_Originally posted by @JacquesCarette in https://github.com/JacquesCarette/Drasil/issues/2436#issuecomment-839853035_

**Comment by balacij** [comment]
I started some preliminary work in the [equationalRealmMK](https://github.com/JacquesCarette/Drasil/compare/equationalRealmMK) branch, but very quickly running into some troubles/big questions.

Assuming we wanted an `EquationalRealm [QDefinition]` data constructor for ModelKinds, ...

1. Would we want to allow it to be instantiated for anything other than TheoryModels? If not, should we potentially have separate ModelKinds ADTs for just TheoryModels?
2. Should each QDefinition in the EquationalRealm be forced to carry the same data? In particular, should each QDefinition in the EquationalRealm have the same UID?
4. If we don't necessarily want all the of QDefinitions to be equivalent (excluding Exprs and definitional Sentences), then when we [grab some information from something that carries a ModelKind](https://github.com/JacquesCarette/Drasil/blob/2b43623ad864ab051db278fd8421629713d22866/code/drasil-theory/Theory/Drasil/ModelKinds.hs#L35), which QDefinition should we grab information from when it's an EquationalRealm? Right now, I set it to be the first element (working on the assumption that each internal QDefinition carries the same information [even though I know they don't, they each, at least, carry different definition Sentences and `relat`s]

However, these questions lead me to think we would actually want some slightly different structure: `EquationalRealm' QuantityDict [(Sentence, Expr)]` where we can use an EquationalRealm to generate QDefinitions (since a QDefinition is currently made up of a QuantityDict, defn :: Sentence, and an Expr [though, technically, it also has a ConceptDomain that isn't used at the moment too])

**Comment by JacquesCarette** [comment]
`NEList`, not `[]`.  We should have it for both `InstanceModel` and `TheoryModel`. At least for now.

So question 2 is much more complicated. It does seem rather wasteful. But it should be easy to create combinators to do variants, so it shouldn't be too bad. They should have different `UID` though. Equivalent is not 'the same'.

It is indeed a non-trivial question to decide which one to choose when grabbing information from such a `ModelKind`. My feeling is that 'we' (i.e. the Drasil program) should not choose. So those instances that would be forced to choose, should not be implemented.

In other words, to use an `EquationalRealm`, a Drasil user will need to explicitly derive an `EquationalModel` by making that choice.

The idea of your "slightly different structure" (modulo `NEList`) is good. I would not use `[(Sentence, Expr)]` explicitly though - there is too high a likelyhood that we'll want to modify the contents of that list. So it should be a record, and the parts should be accesses with lenses, for some amount of future-proofing.

**Comment by JacquesCarette** [comment]
Also: this is thesis fodder, along with multiple similar discussions of late. Keep good notes!

**Comment by balacij** [comment]
Sounds good.

Regarding the `NEList`, I was assuming you meant just a general non-empty list, sorry! Do we have any specific package we want to import for it? I can't find anything exactly named `NEList`, but `base` contains [`NonEmpty`](https://hackage.haskell.org/package/base-4.9.0.0/docs/Data-List-NonEmpty.html) which appears to be a NonEmptyList. Just want to make sure now.

It sounds like we're tending towards the `EquationalRealm' QuantityDict [(Sentence, Expr)]`-based structure for now then (with the `NEList` and record change)? Or maybe we'll need to try out both and see what things look like?

I think that with the `EquationalRealm' QuantityDict [(Sentence, Expr)]`-based version, we might be able to get some amount information from it nicely since we'd primarily be grabbing from QuantityDict and connecting information between the QuantityDict and the Exprs/Definition Sentences together to fill the other gaps.

**Comment by JacquesCarette** [comment]
I might have remembered the Agda name. Yes, I just mean non-empty list, as standard a version as possible.

Yes, that second solution indeed. It does seem to let us access most of the information, just not 'the' definition, because there is no such thing.

**Comment by balacij** [comment]
I just made a rather hacky, but "working", prototype in [equationalRealmMk](https://github.com/JacquesCarette/Drasil/compare/equationalRealmMK) but I'm not quite happy with it yet.

How far do we want to go with EquationalRealms on the first go? Regarding using them in InstanceModels, I think we might even want to disallow it and further restrict them to only TheoryModels.

Do you have any thoughts on this model @JacquesCarette?

**Comment by balacij** [comment]
As a counterpoint to not allowing EquationalRealms in InstanceModels, we currently have (in GamePhysics):

![image](https://user-images.githubusercontent.com/1627302/119193697-ff19dd80-ba4f-11eb-93e1-1c991a3b477e.png)

Though this one has a FIXME related to it:
```
transMotRel :: Relation -- FIXME: add proper equation
transMotRel = sy accj $= deriv (apply1 velj time) time
  $= sy gravitationalAccel `addRe` (apply1 forcej time $/ sy massj)
```

**Comment by JacquesCarette** [comment]
Right, this is going to force us to think harder about "models". @smiths needs to be involved in this discussion too.

Current thinking:
- we need to have all sorts of kinds of models
- models are related to each other; a very important one is 'refinement', where some piece of information that was general becomes more constrained
- at some level, theories might have multiple guises, i.e. have multiple equations that say equivalent things
- a refinement will then pick a single equation to be the one to use from then on in that development (though we need to keep the possibility that a single model can be refined multiple times in the same development - it has to be relational, not functional).

This leads me to think that we'll need more than one `ModelKind` type. [Also, I think we're going to have to "let go" of TM/IM/GD/DD very soon. The distinctions there seem to continue to be too vague. But I don't want to remove them until we've found a reasonable replacement scheme.]

To move forward on this issue, we do need to settle on where `EquationalRealm` can and cannot occur. My guess:
- yes in `TheoryModel`
- yes in `InstanceModel`
- no in `GenDefn`

This does have consequences: it means that one cannot extract a `Definition`, `ConceptDomain` or `ExprRelat` from the first two. Let's analyze a bit more:
- `Definition`: this kind of makes sense as, in our work, each equation expresses some slightly different idea, so the union of them doesn't have a single one. This kind of arguments that models should have their own Definition. More precisely, an `EquationalRealm` should have more than a list of equations, but also a definition that ties them together.
- `ConceptDomain`. This is more interesting. I think that the constructor(s) for an `EquationalRealm` should enforce that all the `ConceptDomain`s of the subparts are equal! Then the `ConceptDomain` could be chosen to be any of them.
- `ExprRelat`: well, that's the whole point, right? There might be more than one. So we need to examine what we do with `ExprRelat`. If it's only for display, that's ok, but if it is used for other things, that's a problem.

**Comment by balacij** [comment]
Thank you!

> This leads me to think that we'll need more than one ModelKind type. [Also, I think we're going to have to "let go" of TM/IM/GD/DD very soon. The distinctions there seem to continue to be too vague. But I don't want to remove them until we've found a reasonable replacement scheme.]

I can see this already applying nicely with our existing ModelKinds not being compatible fully with GenDefns if we choose not to allow EquationalRealms for them.

One issue however is:
![image](https://user-images.githubusercontent.com/1627302/119423915-6ba40f00-bcd2-11eb-9602-c4da6688015c.png)
(from DblPendulum)

Would we want this to be an EquationalRealm? As of right now, it wouldn't be allowed. In which case, we would need to change what it's equation says.

**Comment by smiths** [comment]
I think the TM/GD/IM/DD distinction is valuable from the point of view of documentation.  I don't think the distinction translates directly into the Drasil language though.  I could certainly see GDs being treated the same as TMs inside Drasil.

**Comment by balacij** [comment]
I think we've completed everything that is mentioned on this ticket and from our second last meeting. I think we can close this ticket now.
